### PR TITLE
Quota on Number of RelationshipTemplates

### PR DESCRIPTION
**Pull request by JLSRKonk**
# Readiness checklist

-   [x] I added/updated unit tests.
-   [ ] I added/updated integration tests.
-   [x] I ensured that the PR title is good enough for the changelog.
-   [x] I labeled the PR.

### CODE REVIEW

File: Modules/Relationships/test/Relationships.Application.Tests/Tests/RelationshipTemplates/Commands/CreateRelationshipTemplate/HandlerTests.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        A.CallTo(() => _userContext.GetAddress()).Returns("some-identity-address");
+        A.CallTo(() => _userContext.GetDeviceId()).Returns(DeviceId.Parse("DVCsomedeviceid12345"));
+
+        var handler = CreateHandler(relationshipTemplatesRepository);
+
+        // Act
+        await handler.Handle(command, CancellationToken.None);
+
+        // Assert
+        A.CallTo(() => _eventBus.Publish(A<IntegrationEvent>.That.IsInstanceOf(typeof(RelationshipTemplateCreatedIntegrationEvent)))).MustHaveHappened();
```
Comment (Dannyps) [inline Modules/Relationships/test/Relationships.Application.Tests/Tests/RelationshipTemplates/Commands/CreateRelationshipTemplate/HandlerTests.cs:50]:
```suggestion
        A.CallTo(() => _eventBus.Publish(A<RelationshipTemplateCreatedIntegrationEvent>._).MustHaveHappened();
```

Wouldn't this work?

File: Modules/Quotas/src/Quotas.Infrastructure/Persistence/Repository/RelationshipTemplatesRepository.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+﻿using Backbone.Modules.Quotas.Application.Infrastructure.Persistence.Repository;
+using Backbone.Modules.Quotas.Domain.Aggregates.Relationships;
+using Backbone.Modules.Quotas.Infrastructure.Persistence.Database;
+using Enmeshed.DevelopmentKit.Identity.ValueObjects;
+using Microsoft.EntityFrameworkCore;
+
+namespace Backbone.Modules.Quotas.Infrastructure.Persistence.Repository;
+internal class RelationshipTemplatesRepository : IRelationshipTemplatesRepository
```
Comment (daniel-almeida-konkconsulting) [inline Modules/Quotas/src/Quotas.Infrastructure/Persistence/Repository/RelationshipTemplatesRepository.cs:8]: Why is this class internal?

File: Modules/Quotas/src/Quotas.Infrastructure/Persistence/Repository/RelationshipTemplatesRepository.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+{
+    private readonly IQueryable<RelationshipTemplate> _templatesReadOnly;
```
Comment (daniel-almeida-konkconsulting) [inline Modules/Quotas/src/Quotas.Infrastructure/Persistence/Repository/RelationshipTemplatesRepository.cs:10]: We typically name these variables with readOnly as a prefix and not a suffix (e.g., _readOnlyTemplates)

File: Modules/Quotas/src/Quotas.Application/IntegrationEvents/Incoming/RelationshipTemplateCreated/RelationshipTemplateCreatedIntegrationEventHandler.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+﻿using Backbone.Modules.Quotas.Application.Metrics;
+using Backbone.Modules.Quotas.Domain.Aggregates.Metrics;
+using Enmeshed.BuildingBlocks.Application.Abstractions.Infrastructure.EventBus;
+
+namespace Backbone.Modules.Quotas.Application.IntegrationEvents.Incoming.RelationshipTemplateCreated;
+internal class RelationshipTemplateCreatedIntegrationEventHandler : IIntegrationEventHandler<RelationshipTemplateCreatedIntegrationEvent>
```
Comment (daniel-almeida-konkconsulting) [inline Modules/Quotas/src/Quotas.Application/IntegrationEvents/Incoming/RelationshipTemplateCreated/RelationshipTemplateCreatedIntegrationEventHandler.cs:6]: Why is this class internal?

File: Modules/Quotas/src/Quotas.Application/Metrics/NumberOfRelationshipTemplatesMetricCalculator.cs [hunk 5]
```diff
@@ -0,0 +1,19 @@
+﻿using Backbone.Modules.Quotas.Application.Infrastructure.Persistence.Repository;
+using Backbone.Modules.Quotas.Domain;
+
+namespace Backbone.Modules.Quotas.Application.Metrics;
+internal class NumberOfRelationshipTemplatesMetricCalculator : IMetricCalculator
```
Comment (daniel-almeida-konkconsulting) [inline Modules/Quotas/src/Quotas.Application/Metrics/NumberOfRelationshipTemplatesMetricCalculator.cs:5]: Another internal class.

File: Modules/Quotas/src/Quotas.Infrastructure/Persistence/Repository/RelationshipTemplatesRepository.cs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+    public RelationshipTemplatesRepository(QuotasDbContext dbContext)
+    {
+        _templatesReadOnly = dbContext.RelationshipTemplates.AsNoTracking();
+    }
+
+    public async Task<uint> Count(IdentityAddress createdBy, DateTime createdAtFrom, DateTime createdAtTo, CancellationToken cancellationToken)
+    {
+        var tokensCount = await _templatesReadOnly.CountAsync(t => t.CreatedBy == createdBy.StringValue, cancellationToken);
```
Comment (daniel-almeida-konkconsulting) [inline Modules/Quotas/src/Quotas.Infrastructure/Persistence/Repository/RelationshipTemplatesRepository.cs:19]: Why is this variable named tokensCount?

**Review by daniel-almeida-konkconsulting — CHANGES_REQUESTED** [review]

**Review by daniel-almeida-konkconsulting — DISMISSED** [review]

File: Modules/Quotas/src/Quotas.Infrastructure/Persistence/Database/QuotasDbContext.cs [hunk 7]
```diff
@@ -26,8 +26,11 @@ public class QuotasDbContext : AbstractDbContextBase
 
     public DbSet<Relationship> Relationships { get; set; }
 
+    public DbSet<RelationshipTemplate> RelationshipTemplates { get; set; }
+
     public DbSet<Token> Tokens { get; set; }
 
+
```
Comment (tnotheis) [inline Modules/Quotas/src/Quotas.Infrastructure/Persistence/Database/QuotasDbContext.cs:33]: Remove this empty line

**Review by tnotheis — CHANGES_REQUESTED** [review]
Did you try to create a quota on this metric via the Admin UI?

**Comment by JLSRKonk** [comment]
The creation of the quota through the adminUi works:

image

image

image

**Review by tnotheis — APPROVED** [review]

**Pull request merged**